File: src/main/java/com/github/morningzeng/toolset/utils/ActionUtils.java
package com.github.morningzeng.toolset.utils;

import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.ActionPlaces;
import com.intellij.openapi.actionSystem.ActionPopupMenu;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.DefaultActionGroup;
import org.jetbrains.annotations.NotNull;

import javax.swing.Icon;
import java.awt.Component;

/**
 * @author Morning Zeng
 * @since 2024-06-16
 */
public final class ActionUtils {

    public static AnAction drawerActions(final String text, final String description, final Icon icon, final AnAction... actions) {
        final DefaultActionGroup actionGroup = new DefaultActionGroup(text, description, icon);
        actionGroup.addAll(actions);
        actionGroup.setPopup(true);

        return new AnAction(text, description, icon) {
            @Override
            public void actionPerformed(@NotNull final AnActionEvent e) {
                final ActionPopupMenu popupMenu = ActionManager.getInstance().createActionPopupMenu(ActionPlaces.POPUP, actionGroup);

                final Component component = e.getInputEvent().getComponent();
                final int width = component.getWidth();
                final int height = component.getHeight();
                popupMenu.getComponent().show(component, width, height);
            }
        };
    }

}

File: src/main/kotlin/com/github/morningzeng/toolbox/utils/ActionUtils.kt
package com.github.morningzeng.toolbox.utils

import com.intellij.openapi.actionSystem.*
import javax.swing.Icon

/**
 * @author Morning Zeng
 * @since 2025-05-19
 */
object ActionUtils {

    fun drawerActions(text: String, description: String, icon: Icon, vararg actions: AnAction): AnAction {
        val group = DefaultActionGroup(text, description, icon).apply {
            addAll(*actions)
            isPopup = true
        }
        return object : AnAction(text, description, icon) {
            override fun actionPerformed(e: AnActionEvent) {
                val popupMenu = ActionManager.getInstance().createActionPopupMenu(ActionPlaces.POPUP, group)
                e.inputEvent?.component?.let { popupMenu.component.show(it, it.width, it.height) }
            }
        }
    }
}